File: api/src/test/java/org/pmiops/workbench/db/dao/DatasetServiceTest.java
package org.pmiops.workbench.db.dao;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.pmiops.workbench.db.model.DbStorageEnums.domainToStorage;

import com.google.cloud.bigquery.Field;
import com.google.cloud.bigquery.FieldList;
import com.google.cloud.bigquery.FieldValue;
import com.google.cloud.bigquery.FieldValue.Attribute;
import com.google.cloud.bigquery.FieldValueList;
import com.google.cloud.bigquery.LegacySQLTypeName;
import com.google.cloud.bigquery.QueryJobConfiguration;
import com.google.cloud.bigquery.QueryParameterValue;
import com.google.cloud.bigquery.StandardSQLTypeName;
import com.google.cloud.bigquery.TableResult;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.gson.Gson;
import java.time.Clock;
import java.time.Instant;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import javax.inject.Provider;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.pmiops.workbench.api.BigQueryService;
import org.pmiops.workbench.cdr.ConceptBigQueryService;
import org.pmiops.workbench.cdr.dao.DSLinkingDao;
import org.pmiops.workbench.cohortbuilder.CohortQueryBuilder;
import org.pmiops.workbench.cohorts.CohortService;
import org.pmiops.workbench.conceptset.ConceptSetService;
import org.pmiops.workbench.config.CdrBigQuerySchemaConfigService;
import org.pmiops.workbench.config.WorkbenchConfig;
import org.pmiops.workbench.dataset.DatasetServiceImpl;
import org.pmiops.workbench.dataset.DatasetServiceImpl.QueryAndParameters;
import org.pmiops.workbench.dataset.mapper.DatasetMapper;
import org.pmiops.workbench.dataset.mapper.DatasetMapperImpl;
import org.pmiops.workbench.db.model.DbCohort;
import org.pmiops.workbench.db.model.DbConceptSet;
import org.pmiops.workbench.exceptions.BadRequestException;
import org.pmiops.workbench.model.DatasetRequest;
import org.pmiops.workbench.model.Domain;
import org.pmiops.workbench.model.DomainValuePair;
import org.pmiops.workbench.model.PrePackagedConceptSetEnum;
import org.pmiops.workbench.model.SearchRequest;
import org.pmiops.workbench.test.FakeClock;
import org.pmiops.workbench.test.SearchRequests;
import org.pmiops.workbench.utils.mappers.CommonMappers;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.junit4.SpringRunner;

@RunWith(SpringRunner.class)
@DataJpaTest
public class DatasetServiceTest {

  private static final QueryJobConfiguration QUERY_JOB_CONFIGURATION_1 =
      QueryJobConfiguration.newBuilder(
              "SELECT * FROM person_id from `${projectId}.${datasetId}.person` person")
          .addNamedParameter(
              "foo",
              QueryParameterValue.newBuilder()
                  .setType(StandardSQLTypeName.INT64)
                  .setValue(Long.toString(101L))
                  .build())
          .build();
  private static final ImmutableList<Long> COHORT_IDS = ImmutableList.of(101L, 102L);
  private static final String TEST_CDR_PROJECT_ID = "all-of-us-ehr-dev";
  private static final String TEST_CDR_DATA_SET_ID = "synthetic_cdr20180606";
  private static final String TEST_CDR_TABLE = TEST_CDR_PROJECT_ID + "." + TEST_CDR_DATA_SET_ID;
  private static final Instant NOW = Instant.now();
  private static final FakeClock CLOCK = new FakeClock(NOW, ZoneId.systemDefault());

  @Autowired private BigQueryService bigQueryService;
  @Autowired private CdrBigQuerySchemaConfigService cdrBigQuerySchemaConfigService;
  @Autowired private CohortDao cohortDao;
  @Autowired private ConceptSetDao conceptSetDao;
  @Autowired private ConceptBigQueryService conceptBigQueryService;
  @Autowired private CohortQueryBuilder cohortQueryBuilder;
  @Autowired private DataDictionaryEntryDao dataDictionaryEntryDao;
  @Autowired private DatasetDao datasetDao;
  @Autowired private DSLinkingDao dsLinkingDao;
  @Autowired private DatasetMapper datasetMapper;
  @Autowired private Provider<WorkbenchConfig> workbenchConfigProvider;

  @MockBean private BigQueryService mockBigQueryService;
  @MockBean private CohortDao mockCohortDao;

  private DatasetServiceImpl datasetServiceImpl;

  @TestConfiguration
  @Import({DatasetMapperImpl.class})
  @MockBean({
    CdrBigQuerySchemaConfigService.class,
    CommonMappers.class,
    CohortService.class,
    ConceptBigQueryService.class,
    ConceptSetDao.class,
    ConceptSetService.class,
    CohortQueryBuilder.class,
    DatasetDao.class
  })
  static class Configuration {
    @Bean
    Clock clock() {
      return CLOCK;
    }

    @Bean
    WorkbenchConfig workbenchConfig() {
      WorkbenchConfig workbenchConfig = new WorkbenchConfig();
      workbenchConfig.featureFlags = new WorkbenchConfig.FeatureFlagsConfig();
      return workbenchConfig;
    }
  }

  @Before
  public void setUp() {
    datasetServiceImpl =
        new DatasetServiceImpl(
            bigQueryService,
            cdrBigQuerySchemaConfigService,
            cohortDao,
            conceptBigQueryService,
            conceptSetDao,
            cohortQueryBuilder,
            dataDictionaryEntryDao,
            datasetDao,
            dsLinkingDao,
            datasetMapper,
            CLOCK,
            workbenchConfigProvider);

    final DbCohort cohort = buildSimpleCohort();
    when(cohortDao.findCohortByNameAndWorkspaceId(anyString(), anyLong())).thenReturn(cohort);
    when(cohortQueryBuilder.buildParticipantIdQuery(any())).thenReturn(QUERY_JOB_CONFIGURATION_1);
  }

  private DbCohort buildSimpleCohort() {
    final SearchRequest searchRequest = SearchRequests.males();
    final String cohortCriteria = new Gson().toJson(searchRequest);

    final DbCohort cohortDbModel = new DbCohort();
    cohortDbModel.setCohortId(101L);
    cohortDbModel.setType("foo");
    cohortDbModel.setWorkspaceId(1L);
    cohortDbModel.setCriteria(cohortCriteria);
    return cohortDbModel;
  }

  private DbConceptSet buildConceptSet(long conceptSetId, Domain domain, Set<Long> conceptIds) {
    DbConceptSet result = new DbConceptSet();
    result.setConceptSetId(conceptSetId);
    result.setDomain(domainToStorage(domain));
    result.setConceptIds(conceptIds);
    return result;
  }

  private static DatasetRequest buildEmptyRequest() {
    final DatasetRequest invalidRequest = new DatasetRequest();
    invalidRequest.setDomainValuePairs(Collections.emptyList());
    invalidRequest.setPrePackagedConceptSet(PrePackagedConceptSetEnum.NONE);
    return invalidRequest;
  }

  @Test(expected = BadRequestException.class)
  public void testThrowsForNoCohortOrConcept() {
    final DatasetRequest invalidRequest = buildEmptyRequest();
    invalidRequest.setDomainValuePairs(
        ImmutableList.of(new DomainValuePair().domain(Domain.CONDITION)));
    datasetServiceImpl.domainToBigQueryConfig(invalidRequest);
  }

  @Test
  public void testGetsCohortQueryStringAndCollectsNamedParameters() {
    final DbCohort cohortDbModel = buildSimpleCohort();
    final QueryAndParameters queryAndParameters =
        datasetServiceImpl.getCohortQueryStringAndCollectNamedParameters(cohortDbModel);
    assertThat(queryAndParameters.getQuery()).isNotEmpty();
    assertThat(queryAndParameters.getNamedParameterValues()).isNotEmpty();
  }

  @Test
  public void testRejectsConceptSetListWithNoConcepts() {
    final DbConceptSet conceptSet1 = new DbConceptSet();
    conceptSet1.setDomain((short) Domain.DEVICE.ordinal());
    conceptSet1.setConceptIds(Collections.emptySet());
    final boolean isValid =
        datasetServiceImpl.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
            ImmutableList.of(conceptSet1));
    assertThat(isValid).isFalse();
  }

  @Test
  public void testAcceptsTwoDomainsWithConcepts() {
    final DbConceptSet conceptSet1 = new DbConceptSet();
    conceptSet1.setDomain((short) Domain.DEVICE.ordinal());
    conceptSet1.setConceptIds(ImmutableSet.of(1L, 2L, 3L));

    final DbConceptSet conceptSet2 = new DbConceptSet();
    conceptSet2.setDomain((short) Domain.PERSON.ordinal());
    conceptSet2.setConceptIds(ImmutableSet.of(4L, 5L, 6L));

    final boolean isValid =
        datasetServiceImpl.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
            ImmutableList.of(conceptSet1, conceptSet2));
    assertThat(isValid).isTrue();
  }

  @Test
  public void testRejectsSomeDomainsWithConceptsSomeWithout() {
    final DbConceptSet conceptSet1 = new DbConceptSet();
    conceptSet1.setDomain((short) Domain.DEVICE.ordinal());
    conceptSet1.setConceptIds(ImmutableSet.of(1L, 2L, 3L));

    final DbConceptSet conceptSet2 = new DbConceptSet();
    conceptSet2.setDomain((short) Domain.PERSON.ordinal());
    conceptSet2.setConceptIds(ImmutableSet.of(4L, 5L, 6L));

    final DbConceptSet conceptSet3 = new DbConceptSet();
    conceptSet3.setDomain((short) Domain.DRUG.ordinal());
    conceptSet3.setConceptIds(Collections.emptySet());

    final boolean isValid =
        datasetServiceImpl.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
            ImmutableList.of(conceptSet1, conceptSet2, conceptSet3));
    assertThat(isValid).isFalse();
  }

  @Test
  public void testAcceptsEmptyConceptSetIfDomainIsPopulated() {
    final DbConceptSet conceptSet1 = new DbConceptSet();
    conceptSet1.setDomain((short) Domain.DEVICE.ordinal());
    conceptSet1.setConceptIds(ImmutableSet.of(1L, 2L, 3L));

    final DbConceptSet conceptSet2 = new DbConceptSet();
    conceptSet2.setDomain((short) Domain.PERSON.ordinal());
    conceptSet2.setConceptIds(ImmutableSet.of(4L, 5L, 6L));

    final DbConceptSet conceptSet3 = new DbConceptSet();
    conceptSet3.setDomain((short) Domain.DEVICE.ordinal());
    conceptSet3.setConceptIds(Collections.emptySet());

    final boolean isValid =
        datasetServiceImpl.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
            ImmutableList.of(conceptSet1, conceptSet2, conceptSet3));
    assertThat(isValid).isTrue();
  }

  @Test
  public void testRejectsEmptyConceptSetList() {
    final boolean isValid =
        datasetServiceImpl.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
            Collections.emptyList());
    assertThat(isValid).isFalse();
  }

  @Test
  public void testBuildConceptIdListClause_same() {
    Domain domain1 = Domain.CONDITION;
    DbConceptSet conceptSet1 = buildConceptSet(1L, domain1, ImmutableSet.of(1L, 2L, 3L));
    DbConceptSet conceptSet2 = buildConceptSet(2L, domain1, ImmutableSet.of(4L, 5L, 6L));
    Optional<String> listClauseMaybe =
        datasetServiceImpl.buildConceptIdListClause(
            domain1, ImmutableList.of(conceptSet1, conceptSet2));
    assertThat(listClauseMaybe.map(String::trim).orElse("")).isEqualTo("IN (1, 2, 3, 4, 5, 6)");
  }

  @Test
  public void testBuildConceptIdListClause_differentDomains() {
    DbConceptSet conceptSet1 = buildConceptSet(1L, Domain.CONDITION, ImmutableSet.of(1L, 2L, 3L));
    DbConceptSet conceptSet2 = buildConceptSet(2L, Domain.DRUG, ImmutableSet.of(4L, 5L, 6L));
    Optional<String> listClauseMaybe =
        datasetServiceImpl.buildConceptIdListClause(
            Domain.CONDITION, ImmutableList.of(conceptSet1, conceptSet2));
    assertThat(listClauseMaybe.map(String::trim).orElse("")).isEqualTo("IN (1, 2, 3)");
  }

  @Test
  public void testBuildConceptIdListClause_noClauseForPersonDomain() {
    DbConceptSet conceptSet1 = buildConceptSet(1L, Domain.CONDITION, ImmutableSet.of(1L, 2L, 3L));
    DbConceptSet conceptSet2 = buildConceptSet(2L, Domain.DRUG, ImmutableSet.of(4L, 5L, 6L));
    Optional<String> listClauseMaybe =
        datasetServiceImpl.buildConceptIdListClause(
            Domain.PERSON, ImmutableList.of(conceptSet1, conceptSet2));
    assertThat(listClauseMaybe.isPresent()).isFalse();
  }

  @Test
  public void testcapitalizeFirstCharacterOnly_uppercaseWord() {
    assertThat(DatasetServiceImpl.capitalizeFirstCharacterOnly("QWERTY")).isEqualTo("Qwerty");
  }

  @Test
  public void testcapitalizeFirstCharacterOnly_mixedCaseString() {
    assertThat(DatasetServiceImpl.capitalizeFirstCharacterOnly("aLl YouR baSE"))
        .isEqualTo("All your base");
  }

  @Test
  public void testcapitalizeFirstCharacterOnly_singleLetterStrings() {
    assertThat(DatasetServiceImpl.capitalizeFirstCharacterOnly("a")).isEqualTo("A");
    assertThat(DatasetServiceImpl.capitalizeFirstCharacterOnly("B")).isEqualTo("B");
  }

  @Test
  public void testcapitalizeFirstCharacterOnly_emptyString() {
    assertThat(DatasetServiceImpl.capitalizeFirstCharacterOnly("")).isEqualTo("");
  }

  @Test
  public void testCapitalizeFirstCharacterOnly_emoji() {
    assertThat(DatasetServiceImpl.capitalizeFirstCharacterOnly("\uD83D\uDCAF"))
        .isEqualTo("\uD83D\uDCAF");
    assertThat((DatasetServiceImpl.capitalizeFirstCharacterOnly("マリオに感謝しますが、私たちの王女は別の城にいます")))
        .isEqualTo("マリオに感謝しますが、私たちの王女は別の城にいます");
  }

  @Test
  public void testDomainToBigQueryConfig() {
    mockLinkingTableQuery(ImmutableList.of("FROM `" + TEST_CDR_TABLE + ".person` person"));
    final DatasetRequest datasetRequest =
        new DatasetRequest()
            .conceptSetIds(Collections.emptyList())
            .cohortIds(Collections.emptyList())
            .domainValuePairs(ImmutableList.of(new DomainValuePair()))
            .name("blah")
            .prePackagedConceptSet(PrePackagedConceptSetEnum.NONE)
            .cohortIds(COHORT_IDS)
            .domainValuePairs(
                ImmutableList.of(new DomainValuePair().domain(Domain.PERSON).value("PERSON_ID")));
    final Gson gson = new Gson();
    final String cohortCriteriaJson =
        gson.toJson(
            new SearchRequest()
                .includes(Collections.emptyList())
                .excludes(Collections.emptyList())
                .dataFilters(Collections.emptyList()),
            SearchRequest.class);

    final DbCohort dbCohort = new DbCohort();
    dbCohort.setCriteria(cohortCriteriaJson);
    dbCohort.setCohortId(COHORT_IDS.get(0));

    doReturn(ImmutableList.of(dbCohort)).when(mockCohortDao).findAllByCohortIdIn(anyList());

    final Map<String, QueryJobConfiguration> result =
        datasetServiceImpl.domainToBigQueryConfig(datasetRequest);
    assertThat(result).hasSize(1);
    assertThat(result.get("PERSON").getNamedParameters()).hasSize(1);
    assertThat(result.get("PERSON").getNamedParameters().get("foo_101").getValue())
        .isEqualTo("101");
  }

  private void mockLinkingTableQuery(Collection<String> domainBaseTables) {
    final TableResult tableResultMock = mock(TableResult.class);

    final FieldList schema =
        FieldList.of(
            ImmutableList.of(
                Field.of("OMOP_SQL", LegacySQLTypeName.STRING),
                Field.of("JOIN_VALUE", LegacySQLTypeName.STRING)));

    doReturn(
            domainBaseTables.stream()
                .map(
                    domainBaseTable -> {
                      ArrayList<FieldValue> rows = new ArrayList<>();
                      rows.add(FieldValue.of(Attribute.PRIMITIVE, "PERSON_ID"));
                      rows.add(FieldValue.of(Attribute.PRIMITIVE, domainBaseTable));
                      return FieldValueList.of(rows, schema);
                    })
                .collect(ImmutableList.toImmutableList()))
        .when(tableResultMock)
        .getValues();

    doReturn(tableResultMock).when(mockBigQueryService).executeQuery(any());
  }
}

File: api/src/test/java/org/pmiops/workbench/db/dao/DataSetServiceTest.kt
package org.pmiops.workbench.db.dao

import com.google.common.truth.Truth.assertThat
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyLong
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.`when`
import org.pmiops.workbench.db.model.CommonStorageEnums.domainToStorage

import com.google.cloud.bigquery.QueryJobConfiguration
import com.google.cloud.bigquery.QueryParameterValue
import com.google.cloud.bigquery.StandardSQLTypeName
import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableSet
import com.google.gson.Gson
import java.util.Collections
import java.util.Optional
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.pmiops.workbench.api.BigQueryService
import org.pmiops.workbench.cdr.ConceptBigQueryService
import org.pmiops.workbench.cohortbuilder.CohortQueryBuilder
import org.pmiops.workbench.config.CdrBigQuerySchemaConfigService
import org.pmiops.workbench.db.dao.DataSetServiceImpl.QueryAndParameters
import org.pmiops.workbench.db.model.Cohort
import org.pmiops.workbench.db.model.ConceptSet
import org.pmiops.workbench.exceptions.BadRequestException
import org.pmiops.workbench.model.DataSetRequest
import org.pmiops.workbench.model.Domain
import org.pmiops.workbench.model.SearchRequest
import org.pmiops.workbench.test.SearchRequests
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.context.annotation.Import
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@DataJpaTest
@Import(LiquibaseAutoConfiguration::class)
class DataSetServiceTest {

    @Autowired
    private val bigQueryService: BigQueryService? = null
    @Autowired
    private val cdrBigQuerySchemaConfigService: CdrBigQuerySchemaConfigService? = null
    @Autowired
    private val cohortDao: CohortDao? = null
    @Autowired
    private val conceptSetDao: ConceptSetDao? = null
    @Autowired
    private val conceptBigQueryService: ConceptBigQueryService? = null
    @Autowired
    private val cohortQueryBuilder: CohortQueryBuilder? = null
    @Autowired
    private val dataSetDao: DataSetDao? = null

    private var dataSetServiceImpl: DataSetServiceImpl? = null

    @TestConfiguration
    @MockBean(BigQueryService::class, CdrBigQuerySchemaConfigService::class, CohortDao::class, ConceptBigQueryService::class, ConceptSetDao::class, CohortQueryBuilder::class, DataSetDao::class)
    internal class Configuration

    @Before
    fun setUp() {
        dataSetServiceImpl = DataSetServiceImpl(
                bigQueryService,
                cdrBigQuerySchemaConfigService,
                cohortDao,
                conceptBigQueryService,
                conceptSetDao,
                cohortQueryBuilder,
                dataSetDao)

        val cohort = buildSimpleCohort()
        `when`(cohortDao!!.findCohortByNameAndWorkspaceId(anyString(), anyLong())).thenReturn(cohort)
        `when`(cohortQueryBuilder!!.buildParticipantIdQuery(any<ParticipantCriteria>()))
                .thenReturn(
                        QueryJobConfiguration.newBuilder(
                                "SELECT * FROM person_id from `\${projectId}.\${dataSetId}.person` person")
                                .build())
        `when`(bigQueryService!!.filterBigQueryConfig(any(QueryJobConfiguration::class.java)))
                .thenReturn(QUERY_JOB_CONFIGURATION_1)
    }

    private fun buildSimpleCohort(): Cohort {
        val searchRequest = SearchRequests.males()
        val cohortCriteria = Gson().toJson(searchRequest)

        val cohortDbModel = Cohort()
        cohortDbModel.cohortId = 101L
        cohortDbModel.type = "foo"
        cohortDbModel.workspaceId = 1L
        cohortDbModel.criteria = cohortCriteria
        return cohortDbModel
    }

    private fun buildConceptSet(conceptSetId: Long, domain: Domain, conceptIds: Set<Long>): ConceptSet {
        val result = ConceptSet()
        result.conceptSetId = conceptSetId
        result.domain = domainToStorage(domain)
        result.conceptIds = conceptIds
        return result
    }

    @Test(expected = BadRequestException::class)
    fun testThrowsForNoCohortOrConcept() {
        val invalidRequest = buildEmptyRequest()
        dataSetServiceImpl!!.generateQueryJobConfigurationsByDomainName(invalidRequest)
    }

    @Test
    fun testGetsCohortQueryStringAndCollectsNamedParameters() {
        val cohortDbModel = buildSimpleCohort()
        val queryAndParameters = dataSetServiceImpl!!.getCohortQueryStringAndCollectNamedParameters(cohortDbModel)
        assertThat(queryAndParameters.query).isNotEmpty()
        assertThat(queryAndParameters.namedParameterValues).isNotEmpty()
    }

    @Test
    fun testRejectsConceptSetListWithNoConcepts() {
        val conceptSet1 = ConceptSet()
        conceptSet1.domain = Domain.DEVICE.ordinal()
        conceptSet1.conceptIds = emptySet()
        val isValid = dataSetServiceImpl!!.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
                ImmutableList.of(conceptSet1))
        assertThat(isValid).isFalse()
    }

    @Test
    fun testAcceptsTwoDomainsWithConcepts() {
        val conceptSet1 = ConceptSet()
        conceptSet1.domain = Domain.DEVICE.ordinal()
        conceptSet1.conceptIds = ImmutableSet.of(1L, 2L, 3L)

        val conceptSet2 = ConceptSet()
        conceptSet2.domain = Domain.PERSON.ordinal()
        conceptSet2.conceptIds = ImmutableSet.of(4L, 5L, 6L)

        val isValid = dataSetServiceImpl!!.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
                ImmutableList.of(conceptSet1, conceptSet2))
        assertThat(isValid).isTrue()
    }

    @Test
    fun testRejectsSomeDomainsWithConceptsSomeWithout() {
        val conceptSet1 = ConceptSet()
        conceptSet1.domain = Domain.DEVICE.ordinal()
        conceptSet1.conceptIds = ImmutableSet.of(1L, 2L, 3L)

        val conceptSet2 = ConceptSet()
        conceptSet2.domain = Domain.PERSON.ordinal()
        conceptSet2.conceptIds = ImmutableSet.of(4L, 5L, 6L)

        val conceptSet3 = ConceptSet()
        conceptSet3.domain = Domain.DRUG.ordinal()
        conceptSet3.conceptIds = emptySet()

        val isValid = dataSetServiceImpl!!.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
                ImmutableList.of(conceptSet1, conceptSet2, conceptSet3))
        assertThat(isValid).isFalse()
    }

    @Test
    fun testAcceptsEmptyConceptSetIfDomainIsPopulated() {
        val conceptSet1 = ConceptSet()
        conceptSet1.domain = Domain.DEVICE.ordinal()
        conceptSet1.conceptIds = ImmutableSet.of(1L, 2L, 3L)

        val conceptSet2 = ConceptSet()
        conceptSet2.domain = Domain.PERSON.ordinal()
        conceptSet2.conceptIds = ImmutableSet.of(4L, 5L, 6L)

        val conceptSet3 = ConceptSet()
        conceptSet3.domain = Domain.DEVICE.ordinal()
        conceptSet3.conceptIds = emptySet()

        val isValid = dataSetServiceImpl!!.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
                ImmutableList.of(conceptSet1, conceptSet2, conceptSet3))
        assertThat(isValid).isTrue()
    }

    @Test
    fun testRejectsEmptyConceptSetList() {
        val isValid = dataSetServiceImpl!!.conceptSetSelectionIsNonemptyAndEachDomainHasAtLeastOneConcept(
                emptyList())
        assertThat(isValid).isFalse()
    }

    @Test
    fun testBuildConceptIdListClause_same() {
        val domain1 = Domain.CONDITION
        val conceptSet1 = buildConceptSet(1L, domain1, ImmutableSet.of(1L, 2L, 3L))
        val conceptSet2 = buildConceptSet(2L, domain1, ImmutableSet.of(4L, 5L, 6L))
        val listClauseMaybe = dataSetServiceImpl!!.buildConceptIdListClause(
                domain1, ImmutableList.of(conceptSet1, conceptSet2))
        assertThat(listClauseMaybe.isPresent).isTrue()
        assertThat(listClauseMaybe.get() == "1 2 3 4 5 6")
    }

    @Test
    fun testBuildConceptIdListClause_differentDomains() {
        val conceptSet1 = buildConceptSet(1L, Domain.CONDITION, ImmutableSet.of(1L, 2L, 3L))
        val conceptSet2 = buildConceptSet(2L, Domain.DRUG, ImmutableSet.of(4L, 5L, 6L))
        val listClauseMaybe = dataSetServiceImpl!!.buildConceptIdListClause(
                Domain.CONDITION, ImmutableList.of(conceptSet1, conceptSet2))
        assertThat(listClauseMaybe.isPresent).isTrue()
        assertThat(listClauseMaybe.get() == "1 2 3")
    }

    @Test
    fun testBuildConceptIdListClause_noClauseForPersonDomain() {
        val conceptSet1 = buildConceptSet(1L, Domain.CONDITION, ImmutableSet.of(1L, 2L, 3L))
        val conceptSet2 = buildConceptSet(2L, Domain.DRUG, ImmutableSet.of(4L, 5L, 6L))
        val listClauseMaybe = dataSetServiceImpl!!.buildConceptIdListClause(
                Domain.PERSON, ImmutableList.of(conceptSet1, conceptSet2))
        assertThat(listClauseMaybe.isPresent).isFalse()
    }

    @Test
    fun testcapitalizeFirstCharacterOnly_uppercaseWord() {
        assertThat(DataSetServiceImpl.capitalizeFirstCharacterOnly("QWERTY")).isEqualTo("Qwerty")
    }

    @Test
    fun testcapitalizeFirstCharacterOnly_mixedCaseString() {
        assertThat(DataSetServiceImpl.capitalizeFirstCharacterOnly("aLl YouR baSE"))
                .isEqualTo("All your base")
    }

    @Test
    fun testcapitalizeFirstCharacterOnly_singleLetterStrings() {
        assertThat(DataSetServiceImpl.capitalizeFirstCharacterOnly("a")).isEqualTo("A")
        assertThat(DataSetServiceImpl.capitalizeFirstCharacterOnly("B")).isEqualTo("B")
    }

    @Test
    fun testcapitalizeFirstCharacterOnly_emptyString() {
        assertThat(DataSetServiceImpl.capitalizeFirstCharacterOnly("")).isEqualTo("")
    }

    @Test
    fun testCapitalizeFirstCharacterOnly_emoji() {
        assertThat(DataSetServiceImpl.capitalizeFirstCharacterOnly("\uD83D\uDCAF"))
                .isEqualTo("\uD83D\uDCAF")
        assertThat(DataSetServiceImpl.capitalizeFirstCharacterOnly("マリオに感謝しますが、私たちの王女は別の城にいます"))
                .isEqualTo("マリオに感謝しますが、私たちの王女は別の城にいます")
    }

    companion object {
        private val QUERY_JOB_CONFIGURATION_1 = QueryJobConfiguration.newBuilder(
                "SELECT * FROM person_id from `\${projectId}.\${dataSetId}.person` person")
                .addNamedParameter(
                        "foo",
                        QueryParameterValue.newBuilder()
                                .setType(StandardSQLTypeName.INT64)
                                .setValue(java.lang.Long.toString(101L))
                                .build())
                .build()

        private fun buildEmptyRequest(): DataSetRequest {
            val invalidRequest = DataSetRequest()
            invalidRequest.setDomainValuePairs(emptyList<T>())
            return invalidRequest
        }
    }
}
